src/pkg/user/manager.go

```diff
@@ -75,20 +76,33 @@ func (m *manager) Delete(ctx context.Context, id int) error {
 	return m.dao.Update(ctx, u, "username", "email", "deleted")
 }
 
-func (m *manager) VerifyLocalPassword(ctx context.Context, username, password string) (bool, error) {
-	u, err := m.GetByName(ctx, username)
+func (m *manager) MatchLocalPassword(ctx context.Context, usernameOrEmail, password string) (*models.User, error) {
+	l, err := m.List(ctx, q.New(q.KeyWords{"email": usernameOrEmail}))
 	if err != nil {
-		return false, err
+		return nil, err
+	}
+	u, err := m.GetByName(ctx, usernameOrEmail)
+	if err != nil {
+		return nil, err
 	}
-	return utils.Encrypt(password, u.Salt, u.PasswordVersion) == u.Password, nil
+	l = append(l, u)
+	for _, entry := range l {
+		if utils.Encrypt(password, entry.Salt, entry.PasswordVersion) == entry.Password {
+			return entry, nil
+		}
+	}
+	return nil, nil
```

Comment:
There is a special char `@` in the email, it is invalid in the username.

Reply:
But this is for querying?
I don't recall how the decision was made, but this is for compatibility with `LoginByDb` in `common/dao` pkg:
https://github.com/goharbor/harbor/blob/release-2.2.0/src/common/dao/user.go#L81

Reply:
There are two db queries in this method `MatchLocalPassword `, only one db query in `LoginByDb`.

And when `usernameOrEmail` is a email, `u, err := m.GetByName(ctx, usernameOrEmail)` will must get a not found error.

Reply:
I have added a `FilterBy...` func to user model.